src/databricks/labs/ucx/progress/install.py

```diff
@@ -0,0 +1,75 @@
+import datetime as dt
+import logging
+from dataclasses import dataclass
+
+from databricks.labs.lsql.backends import Dataclass, SqlBackend
+from databricks.sdk.errors import InternalError
+from databricks.sdk.retries import retried
+
+from databricks.labs.ucx.framework.utils import escape_sql_identifier
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, kw_only=True)
+class HistoricalRecord:
+    workspace_id: int
+    """The identifier of the workspace where this record was generated."""
+
+    run_id: int
+    """An identifier of the workflow run that generated this record."""
+
+    snapshot_id: int
+    """An identifier that is unique to the records produced for a given snapshot."""
+
+    run_start_time: dt.datetime
+    """When this record was generated."""
+
+    object_type: str
+    """The inventory table for which this record was generated."""
+
+    object_type_version: int
```

Comment:
we can 
```
version_map[tuple[int,int,int], int] = {
   (0,37,0): 1,
   (0,39,0): 2,
   (0,52,0): 3,
}
```

Reply:
Does this handle future versions, including later versions installed in other workspaces attached to the same metastore?

If we (a reader) are on version `0,54,0` then we'll assume version 3 (which is fine) but break if `0,55,0` introduced version 4 because that's indistinguishable to us: we will treat it was 3 (and fail) because only `0,55,0` or later knows version 4 exists.

Just encoding the integer avoids this because we can detect a future record (and that it cannot be decoded by the current version).

Reply:
actually, better option would be:

```
migrate_after_versions[tuple[int,int,int], Callable[[dict[str,str]], dict[str,str]] = {
   (0,37,0): lambda x: x | {'foo': "bar"},
   (0,39,0): remove_x_column,
   (0,52,0): convert_some_datatype,
}
```

we sort by version tuple and get a diff between `from databricks.labs.ucx.__about__ import __version__` and apply those. if we get a version that we don't understand - we warn and skip.

Reply:
Just to aid my own understanding, and try to work through ambiguities, as far as I can tell there are two ways of interpreting a version tied to the UCX release:

 - The records contain the _first version_ that understands the serialised form. That is, version 0.37.0, 0.38.0, etc. write 0.37.0 in the record, until version 0.39.0 at which point version 0.39.0, 0.40, etc. all start writing 0.39.0. This continues until 0.52.0, etc.

   Under this protocol we can detect future/unknown versions because they're higher than `__about__.__version__`. We can also read past versions. It does however have a fairly large caveat: for all data classes we need to know the earliest version at which the 'current' version was introduced. This needs to be known in advance of the software release while we're writing and testing it and hard-coded somewhere. (We also need something special during development to ensure that we can read our own records.)

 - The records contain the _actual version_ that wrote the record.

   Under this protocol we can detect past versions, but not handle future/unknown versions because there's no way to know if a future release introduced a new version in some way. For safety we could just always ignore (and warn on) versions later than `__about__.__version__`. The main caveat here relates to when a downgrade happens. Mixed installs with a shared metastore could end up quite noisy (if not filtering out records from other workspaces).

@nfx: Is this also your understanding?

Reply:
> The records contain the actual version that wrote the record.

this is the intended way. let's store the version as `ucx_version: list[int]` and source it from `__about__.__version__` and use `databricks.sdk.mixins.compute.SemVer.parse` to parse it.

Reply:
That's clear, thanks.